src/drivers/scsi/scsi.c

```diff
@@ -22,81 +22,84 @@ static inline struct usb_mass *scsi2mass(struct scsi_dev *dev) {
 	return member_cast_out(dev, struct usb_mass, scsi_dev);
 }
 
-static int scsi_cmd(struct scsi_dev *sdev, void *cmd, size_t cmd_len,
-		void *data, size_t data_len) {
+int scsi_do_cmd(struct scsi_dev *sdev, struct scsi_cmd *scmd) {
 	struct usb_mass *mass = scsi2mass(sdev);
-	struct scsi_cmd *scmd = cmd;
 	enum usb_direction usb_dir;
 
 	if (!sdev->attached) {
 		return -ENODEV;
 	}
+	assert(scmd->scmd_len <= SCSI_CMD_MAX_LEN);
+
 	if (scmd->scmd_opcode == SCSI_CMD_OPCODE_WRITE10) {
 		usb_dir = USB_DIRECTION_OUT;
 	} else {
 		usb_dir = USB_DIRECTION_IN;
 	}
+	log_debug("opcode 0x%x, usb_dir(%d), len %d",
+		scmd->scmd_opcode, usb_dir, scmd->scmd_len);
 
-	log_debug("opc (0x%x) cmd_len(%d), usb_dir(%d), data_len(%d)",
-		    scmd->scmd_opcode, cmd_len, usb_dir, data_len);
-	
-	return usb_ms_transfer(mass->usb_dev, cmd, cmd_len, usb_dir, data, data_len);
+	return usb_ms_transfer(mass->usb_dev, scmd->scmd_buf, scmd->scmd_len,
+		       usb_dir, scmd->sdata_buf, scmd->sdata_len);
 }
 
-int scsi_do_cmd(struct scsi_dev *dev, struct scsi_cmd *cmd) {
-	uint8_t scmd[SCSI_CMD_MAX_LEN];
-
-	log_debug("opcode 0x%x, len %d", cmd->scmd_opcode, cmd->scmd_len);
-
-	assert(cmd->scmd_len <= SCSI_CMD_MAX_LEN);
-	memset(scmd + 1, 0, cmd->scmd_len - 1);
-	scmd[0] = cmd->scmd_opcode;
-
-	if (cmd->scmd_fixup) {
-		cmd->scmd_fixup(scmd, dev, cmd);
-	}
+static int scsi_send_inquiry(struct scsi_dev *dev) {
+	struct scsi_cmd cmd;
```

Comment:
Looks like `cmd.scmd_opcode` is not set in this function, but it's read in `scsi_do_cmd()`. Set it to zero to make sure there's no garbage there?

Reply:
yes, I will just remove this field, it's not used.

Reply:
Fixed.